Improve segment proofreading in 3D viewport

philippotto opened the pull request:
- allows to activate segment via mesh in 3D viewport
- highlights supervoxels on hover
- allows merging/splitting in 3d viewport by interacting with the agglomerates / supervoxels

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- [create new annotation](https://bettermeshcontextmenu.webknossos.xyz/datasets/sample_organization/l4dense_motta_et_al_dev/createExplorative/hybrid?autoFallbackLayer=true)
- select a hdf5 mapping
- select proofreading tool and load some meshes
- try to split and merge them in the 3D viewport

### TODOs:
- [x] backend:
  - [x] as an alternative to positions, also accept unmappedSegmentIds for split/merge related operations
  - [x] implement API that maps from segmentId to a segmentPosition
- [x] frontend
  - [x] integrate the above ino the context menu in the 3D viewport
  - [x] don't show the proofreading cross in the data viewports if the user activated a supervoxel in the 3D viewport
  - [x] bugs
    - [x] loading the same mesh twice breaks the hovering effect for unknown reasons
    - [X] after proofread operation, selected super voxel is not highlighted anymore (but stored as active)
    - [x] when creating a new annotation, a new active segment id is chosen. when switching to proofreading, one can try to merge with that segment id. however, it doesn't exist anywhere.
    - ~[ ] selecting one partner in the 3D viewport and the other in a data viewport (or vice versa) doesn't work~ the user is told so now. hopefully can be improved with #6569 

### Issues:
- contributes to #7619 

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

philippotto commented:
@fm3 When you have some time, it would be great if you could make the `splitAgglomerate` and `mergeAgglomerate` compatible with passing the unmapped segment ids. The same basically for the `/agglomerateGraphMinCut` and `/getNeighborsForAgglomerateNode` route :pray:

fm3 commented:
@philippotto I’ve done that now, please give it a try! The new params are called `segmentId1`, `segmentId2`, `segmentId`

I assume that the frontend now always has the segment ids, right? Applying the update actions is backwards compatible, but do I need to build the same for the mincut/neighbors routes?

fm3 commented:
> For the mincut/neighbors route in non-3d-viewport, we don't have the unmapped id yet, but we can do another request to get these (like we do it for split/merge). These round trips will then be eliminated with the magic mapping soon.

Right, I wasn’t sure about that. If it’s complicated to do these roundtrips, I can also make the routes accept both variants

normanrz reviewed (approved):
Backend looks good :+1:

MichaelBuessemeyer reviewed (changes requested):
Thank you for this awesome PR :tada:

Here are my review and my comments on these changes.

In the TODOs for this PR the following is listed: 

> implement API that maps from segmentId to a segmentPosition

In my understanding wk already knows where a segment is via the segment stats, correct?  If that's the case, doesnt the frontend already know this when loading a segment via the segment stats? Or is this part of the segments stats not sent to the frontend and thus the backend needs to explicitly serve this information as it is needed?

While reading the code, I am not completely familiar with what each line does but I could not find any mistakes. Besides that, I stumbled upon two (three) cases of duplicate naming for the same concept / object / ...
Would it be possible to have consistent name in these cases? Imo that would help a lot understanding the code.

1. I think it was in the `mesh_saga.ts` where `Chunck` refers to a mesh of an unmapped segment while the file calls the underlying segment something like `unmappedSegment`. Imo it would be nicer to call the mesh `unmappedSegmentMesh` or something similar as it makes the relationship between the `unmappedSegment` and the "Chunk" clearer.

2. Furthermore, throughout the codebase we have the name `agglomerateId` and `mappedSegmentId` for the exact same data / object / think. The same goes for `segmentId` and `umappedSegmentId` (if I understood the code correctly). Imo it would be great to have a single consistent name here. But I can also understand that in some parts of the code one name makes more sense than the other. Do you think it would still be possible to have consistent naming while keeping the code within functions in different "logical contexts" understandable?

Besides this, I did not find anything in the code :rocket:.

I have one more question before testing:
Did I understand proofreading actions correctly that the underlying work / changes to the mapping is solely done by the backend as the frontend only sends update actions to the backend to modify the editable mapping, then awaits for the save queue and after that reloads the modified buckets & meshes to ensure an updated state? Thus, the frontends part is only collecting the required information for the proofreading action, sending it to the backend and then reloading?
If that's correct, I think I understood the code correctly (on a relative high level view)

And one final thing:
What about the bug that occurred during showcasing yesterday? Did you already find and fix it or is this still an open TODO? :)

MichaelBuessemeyer commented:
Here is my testing feedback / observations:

Imo it would be nice to select a supervoxel via `left click`, or `shift + left click` or `ctrl + left click` and not only via the context menu :). <- Just my first UX impression. I accidentally deleted a mesh multiple times while looking for a shortcut :sweat_smile: (via `ctrl + left click`)

I am able to reach an inconsistent state with the super voxel highlighting via the following steps (nothing serious):
1. Select a super voxel in the 3d viewport via the context menu.
2. Split it from all neighbours
3. The supervoxel is still highlighted as active, although when opening the context menu, it allows me to select the supervoxel again

When I select a supervoxel and then open the context menu, I can select the option "Split from active segment" which results in an error (pretty obvious). Is it possible to disable this menu item in this case?


When I select a supervoxel in the 2d viewport and then go to a data viewport and select to merge the highlighted segment, the merging is done. Is that correct / expected? Afaik you told me that when selecting in one of the two (data viewport / 3d viewport) and then merging in the other one, should result in missing information and thus an error.

> What about the bug that occurred during showcasing yesterday? Did you already find and fix it or is this still an open TODO? :)

At least, I cannot reproduce this anymore :-1:

philippotto commented on frontend/javascripts/oxalis/controller/segment_mesh_controller.ts:
Yes, with one refinement:

> Segment Meshes are now rendered by chunks as they are no longer merged together

They **can** be merged (when not in proofreading mode), but still then there is a Group which contains one (merged) geometry per chunk.

philippotto commented on frontend/javascripts/oxalis/model/sagas/proofread_saga.ts:
```diff
@@ -563,21 +622,80 @@ function* handleProofreadMergeOrMinCut(action: Action) {
+      action.segmentId == null
+    ) {
+      Toast.warning(MISSING_INFORMATION_WARNING);
+      console.log("Some fields were null:", {
+        agglomerateId: action.agglomerateId,
+        activeCellId,
+        activeUnmappedSegmentId,
+        segmentId: action.segmentId,
+      });
```
>  Did you intentionally leave this in the code to enable easier debugging this feature in production

Yes, this was my intention. This code will change anyway with the magic mappings and inbetween it seemed like a good idea to make debugging easier for us in production.

philippotto commented on frontend/javascripts/oxalis/model/sagas/proofread_saga.ts:
```diff
@@ -520,6 +571,13 @@ function* performCutFromNeighbors(
 
   for (const edge of edgesToRemove) {
     if (sourceTree) {
+      if (edge.position2 == null) {
+        Toast.warning("Could not perform cut from neighbors. See console for more details.");
+        console.warn(
+          "segmentPosition is not available even though a tree was passed to performCutFromNeighbors.",
+        );
+        return { didCancel: true };
+      }
       const result = getDeleteEdgeActionForEdgePositions(sourceTree, edge);
```
I started writing this comment about line 523, but maybe you meant the added check. I'll explain both:

Line 581:
This code is executed when cutting a supervoxel from all its neighbors given a skeleton tree ("agglomerate tree"). In this case, the backend gives us two positions each edge that should be deleted in the skeleton. Usually, this should simply work, but if there is another bug the look up of the edges may fail. We don't want to crash WK completely which is why we catch this case. Within `getDeleteEdgeActionForEdgePositions` an error toast will be shown.

Lines 574-580:
The code is executed in the same scenario as above. This means we have a skeleton tree at hand. In that case, there should also be a not-null `position2`. However, the typing doesn't express this correlation. Therefore, there is this check. I'll add an comment to the code.

philippotto commented:
> Furthermore, throughout the codebase we have the name agglomerateId and mappedSegmentId for the exact same data / object / think. The same goes for segmentId and umappedSegmentId (if I understood the code correctly). Imo it would be great to have a single consistent name here. But I can also understand that in some parts of the code one name makes more sense than the other. Do you think it would still be possible to have consistent naming while keeping the code within functions in different "logical contexts" understandable?

Unfortunately, I'm not sure how to do this well. As an example, the store has a property for the active cell id (should be renamed to active segment id but this is another issue ^^). If there is a mapping, this refers to the mapped id, which could also be called agglomerate id. However, if there is no mapping, "agglomerateId" is highly misleading..

philippotto commented:
> I think it was in the mesh_saga.ts where Chunck refers to a mesh of an unmapped segment while the file calls the underlying segment something like unmappedSegment. Imo it would be nicer to call the mesh unmappedSegmentMesh or something similar as it makes the relationship between the unmappedSegment and the "Chunk" clearer.

Hmm, similar to the other thread above, I'm not sure how to do this well. I wouldn't want to use `unmappedSegmentMesh`, because (a) it's a confusing term for the frequent case where no mappings exist in the first place and (b) when a mapping exists, the back-end maps the individual mesh chunks to the requested id and then returns these chunks. So, they *are* mapped in a sense. I'm not very happy about it, either, but I'm having trouble to see how to it can be improved easily..

> I have one more question before testing:
Did I understand proofreading actions correctly that the underlying work / changes to the mapping is solely done by the backend as the frontend only sends update actions to the backend to modify the editable mapping, then awaits for the save queue and after that reloads the modified buckets & meshes to ensure an updated state? Thus, the frontends part is only collecting the required information for the proofreading action, sending it to the backend and then reloading?
If that's correct, I think I understood the code correctly (on a relative high level view)

Yes, this is correct :) With the upcoming magic mapping pr, this is going to change, though. Then, the frontend will apply the mapping locally (and maybe also update the mapping locally).

philippotto commented:
@MichaelBuessemeyer Thank you for your thorough testing and feedback :+1: I incorporated most of your feedback, fixed the hiccups you noticed and did some performance optimizations. Please have another look and feel free to test again :)

MichaelBuessemeyer reviewed (changes requested):
Thanks for incorporating the feedback :pray: 

> Unfortunately, I'm not sure how to do this well. As an example, the store has a property for the active cell id (should be renamed to active segment id but this is another issue ^^). If there is a mapping, this refers to the mapped id, which could also be called agglomerate id. However, if there is no mapping, "agglomerateId" is highly misleading..

> Hmm, similar to the other thread above, I'm not sure how to do this well. I wouldn't want to use unmappedSegmentMesh, because (a) it's a confusing term for the frequent case where no mappings exist in the first place and (b) when a mapping exists, the back-end maps the individual mesh chunks to the requested id and then returns these chunks. So, they are mapped in a sense. I'm not very happy about it, either, but I'm having trouble to see how to it can be improved easily..

I can see your valid points :/. It make more sense to keep it the way it is :+1: 

During testing I encountered the following unexpected behaviours:
- When I select the `Split from all neighboring segments` action, I get a `A precomputed mesh could not be loaded for this segment. More information was printed to the browser's console.` warning toast. All segments and buckets are reloaded as expected but this warning toast should not occur imo.


Moreover, during testing I came up with another potential useful feature: When I hover a segment in the data viewports the (mapped) segment id is displayed in the footer bar. But if I hover a supervoxel its mapped segment is not displayed in the footer bar. Would it be possible to add this / create a new issue for this?
In particular I needed this use case to check whether two meshes belong to the same segment or not. This was pretty difficult as their color was nearly identical (while one of the was hovered). To check this, I needed to find the segment in the data viewport and hover them there. Imo supporting 3d viewports here as well would help out a lot in this case :)

Thanks all I could find :tada: 

Btw: Thanks for getting rid of unnecessary rerendering of components once the flycam changed :rocket:

MichaelBuessemeyer commented:
> Can you reproduce this reliably? If so, could you share the steps for reproduction? :)

Oh sorry, for me this behaviour occurred every time I used "Split from all neighboring segments" thus I thought that no explicit steps are needed.

I used the following steps:
- load https://bettermeshcontextmenu.webknossos.xyz/annotations/66310e57010000ad00313c3d#2850,4318,1770,0,1
- load segment "Segment 102775"
- in the 3d viewport somewhere at the nucleus of the segment / cell use the "Split from all neighboring segments"  option from the context menu. Then the warning should pop up

philippotto commented:
Thanks for the repro steps :pray: Indeed, it happens quite frequently on the dev instance for me, too.

@fm3 The error is coming from the back-end (the front-end might still be at fault for sending weird inputs) and I'm not sure whether I interpet it correctly.

The frontend uses the `/meshes/chunks` route and sends a payload à la:

```
{
  "meshFile": "meshfile_4-4-2",
  "segmentId": 7571522
}
```

Error:
```
[
  {
    "error": "Failed to load chunk list for segment {0} from mesh file “{1}”"
  },
  {
    "chain": "[Server Time 2024-05-02 08:04]  <~ zero chunks"
  }
]
```

The string interpolation doesn't seem to work, but I guess the backend simply cannot find chunks for id 7571522. Probably because that supervoxel doesn't exist in the editable mapping, anymore?

philippotto commented:
> When I select the Split from all neighboring segments action, I get a A precomputed mesh could not be loaded for this segment. More information was printed to the browser's console. warning toast. All segments and buckets are reloaded as expected but this warning toast should not occur imo.

As discussed [here](https://scm.slack.com/archives/C5AKLAV0B/p1714640367532199), this issue is unrelated to this PR and even occurs in view-only mode. I created [this issue](https://github.com/scalableminds/webknossos/issues/7778) for the problem.

@MichaelBuessemeyer It would be great if you could have a final look at my latest commits. I'd love to merge the PR today unless someone vetoes :)

MichaelBuessemeyer commented on frontend/javascripts/messages.tsx:
```diff
@@ -234,8 +234,8 @@ instead. Only enable this option if you understand its effect. All layers will n
     "There is no mesh for the active segment id available to download.",
-  "tracing.mesh_listing_failed":
-    "A precomputed mesh could not be loaded for this segment. More information was printed to the browser's console.",
+  "tracing.mesh_listing_failed": (segmentId: number) =>
+    `A precomputed mesh could not be loaded for segment ${segmentId}. You may want to use ad-hoc meshing instead. More information was printed to the browser's console.`,
```
Here you change the message from `string` to `function`

MichaelBuessemeyer commented on frontend/javascripts/oxalis/model/sagas/mesh_saga.ts:
```diff
@@ -792,7 +811,11 @@ function* loadPrecomputedMeshForSegmentId(
     loadingOrder = chunkDescriptors.loadingOrder;
   } catch (exception) {
     Toast.warning(messages["tracing.mesh_listing_failed"]);
```
This Toast call needs to be adjusted as the message now is a function:
```suggestion
    Toast.warning(messages["tracing.mesh_listing_failed"], id);
```

I just checked the code base: This should be the only occurrence of `messages["tracing.mesh_listing_failed"]`.

MichaelBuessemeyer replied:
Sorry, this should be correct:
```suggestion
    Toast.warning(messages["tracing.mesh_listing_failed"](id));
```

MichaelBuessemeyer reviewed (approved):
Thanks for investigating the bug I found :pray: as well as making it possible to deselect a super voxel via `ctrl + click`.

During checking your changes I found a little bug in them: 

You redefined the `tracing-mesh_listing_failed` message in `messages.tsx` to a function but did not adjust the case where this message is used. Please see my indicating in line comments.
I was able to confirm the bug in the dev instance -> the displayed toast is empty:
![image](https://github.com/scalableminds/webknossos/assets/39529669/b6cdb551-f433-4320-bafa-bea3f2b94724)

That's all :tada:.

After this little fix, this PR is ready to go :runner:

philippotto commented:
> During checking your changes I found a little bug in them:

Argh, you are right. TypeScript should have caught this, too, but didn't :cry: Maybe because a function can be a react component and the toast is able to render this. Still... I fixed it now!
